File: Python/3-SlidingWindow/Problems/3-Permutation-in-String/Permutation-in-String.py
'''
Sliding Window: Permutation in String

Problem Statement:
Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.
'''

from collections import Counter

# 1. Brute Force (O(n! * m))
def check_permutation_brute_force(s1, s2):
    """
    Checks if s2 contains a permutation of s1 using a brute-force approach.
    It generates all permutations of s1 and checks if any of them are a substring of s2.
    This is very inefficient for larger strings.

    Args:
        s1 (str): The shorter string.
        s2 (str): The longer string.

    Returns:
        bool: True if s2 contains a permutation of s1, False otherwise.
    """
    from itertools import permutations  # Import the permutations function.
    perms = set("".join(p) for p in permutations(s1))  # Generate all permutations of s1 and store them in a set.
    # "".join(p) converts a permutation tuple to a string.  The set avoids duplicate permutations.
    for i in range(len(s2) - len(s1) + 1):  # Iterate through all possible starting positions of a substring of length s1 in s2.
        if s2[i:i+len(s1)] in perms:  # Check if the current substring of s2 is one of the permutations of s1.
            return True  # If a permutation is found, return True.
    return False  # If no permutation is found, return False.

# 2. Sorting Approach (O(m log m + n log n))
def check_permutation_sorting(s1, s2):
    """
    Checks if s2 contains a permutation of s1 by sorting both s1 and substrings of s2.
    If a sorted substring of s2 is equal to the sorted s1, then s2 contains a permutation of s1.

    Args:
        s1 (str): The shorter string.
        s2 (str): The longer string.

    Returns:
        bool: True if s2 contains a permutation of s1, False otherwise.
    """
    s1_sorted = sorted(s1)  # Sort the characters in s1.
    for i in range(len(s2) - len(s1) + 1):  # Iterate through all possible starting positions of a substring of length s1 in s2.
        if sorted(s2[i:i+len(s1)]) == s1_sorted:  # Sort the characters in the current substring of s2 and compare it with the sorted s1.
            return True  # If they are equal, return True.
    return False  # If no matching substring is found, return False.

# 3. Sliding Window with HashMap (O(n + m))
def check_permutation_sliding_window(s1, s2):
    """
    Checks if s2 contains a permutation of s1 using a sliding window and HashMaps (Counters).
    It maintains a character count of s1 and a character count of the current window in s2.

    Args:
        s1 (str): The shorter string.
        s2 (str): The longer string.

    Returns:
        bool: True if s2 contains a permutation of s1, False otherwise.
    """
    s1_count = Counter(s1)  # Create a Counter (HashMap) for the character counts in s1.
    window_count = Counter(s2[:len(s1)-1])  # Create a Counter for the character counts in the initial window of s2 (of size len(s1)-1).

    for i in range(len(s1)-1, len(s2)):  # Iterate through s2, starting from the index where the window has size len(s1).
        window_count[s2[i]] += 1  # Add the new character entering the window.
        if window_count == s1_count:  # If the character counts in the window match the character counts in s1,
            return True  # then a permutation is found.
        window_count[s2[i - len(s1) + 1]] -= 1  # Remove the character leaving the window.
        if window_count[s2[i - len(s1) + 1]] == 0:  # If the count of a character becomes 0,
            del window_count[s2[i - len(s1) + 1]]  # remove it from the Counter.

    return False  # If no permutation is found, return False.

# 4. Optimized Sliding Window with Fixed Array (O(n))
def check_permutation_optimized(s1, s2):
    """
    Checks if s2 contains a permutation of s1 using a sliding window and fixed-size arrays for character counts.
    This optimization is possible because the number of distinct characters is limited (e.g., 26 for lowercase English).

    Args:
        s1 (str): The shorter string.
        s2 (str): The longer string.

    Returns:
        bool: True if s2 contains a permutation of s1, False otherwise.
    """
    if len(s1) > len(s2):  # If s1 is longer than s2, s2 cannot contain a permutation of s1.
        return False

    s1_map = [0] * 26  # Create an array to store character counts for s1 (assuming lowercase English letters).
    s2_map = [0] * 26  # Create an array to store character counts for the sliding window in s2.
    a_ord = ord('a')  # Get the ASCII value of 'a' for efficient character index calculation.

    for i in range(len(s1)):  # Calculate character counts for s1 and the initial window of s2.
        s1_map[ord(s1[i]) - a_ord] += 1  # Increment the count for the corresponding character in s1_map.
        s2_map[ord(s2[i]) - a_ord] += 1  # Increment the count for the corresponding character in s2_map.

    matches = sum([s1_map[i] == s2_map[i] for i in range(26)])  # Count the number of characters with matching counts.

    for i in range(len(s1), len(s2)):  # Slide the window through s2.
        if matches == 26:  # If all 26 characters have matching counts, a permutation is found.
            return True

        left_char = ord(s2[i - len(s1)]) - a_ord  # Get the character leaving the window.
        right_char = ord(s2[i]) - a_ord  # Get the character entering the window.

        s2_map[right_char] += 1  # Add the new character to the window's count.
        if s1_map[right_char] == s2_map[right_char]:  # If the count now matches s1's count,
            matches += 1  # increment matches.
        elif s1_map[right_char] + 1 == s2_map[right_char]:  # If the count was one less than s1's count,
            matches -= 1  # decrement matches.

        s2_map[left_char] -= 1  # Remove the character leaving the window.
        if s1_map[left_char] == s2_map[left_char]:  # If the count now matches s1's count,
            matches += 1  # increment matches.
        elif s1_map[left_char] - 1 == s2_map[left_char]:  # If the count was one more than s1's count,
            matches -= 1  # decrement matches.

    return matches == 26  # After the loop, check if all characters have matching counts.

# 5. Two Pointers with Character Frequency (O(n))
def check_permutation_two_pointers(s1, s2):
    """
    Checks if s2 contains a permutation of s1 using two pointers and a character frequency counter.
    It maintains a sliding window in s2 and compares the character frequencies in the window with those in s1.

    Args:
        s1 (str): The shorter string.
        s2 (str): The longer string.

    Returns:
        bool: True if s2 contains a permutation of s1, False otherwise.
    """
    target_freq = Counter(s1)  # Create a Counter for the character frequencies in s1.
    window_freq = Counter()  # Create a Counter for the character frequencies in the current window of s2.

    left = 0  # Initialize the left pointer of the sliding window.
    for right in range(len(s2)):  # Iterate through s2 using the right pointer.
        window_freq[s2[right]] += 1  # Add the current character to the window's frequency count.
        if right - left + 1 > len(s1):  # If the window size exceeds the length of s1,
            window_freq[s2[left]] -= 1  # remove the leftmost character from the window.
            if window_freq[s2[left]] == 0:  # If the frequency of the leftmost character becomes 0,
                del window_freq[s2[left]]  # remove it from the Counter.
            left += 1  # Move the left pointer to shrink the window.

        if window_freq == target_freq:  # If the character frequencies in the window match those in s1,
            return True  # a permutation is found.
    return False  # If no permutation is found, return False.

# Test Cases
s1 = "ab"
s2 = "eidbaooo"
print("Brute Force:", check_permutation_brute_force(s1, s2))
print("Sorting Approach:", check_permutation_sorting(s1, s2))
print("Sliding Window with HashMap:", check_permutation_sliding_window(s1, s2))
print("Optimized Sliding Window:", check_permutation_optimized(s1, s2))
print("Two Pointers with Character Frequency:", check_permutation_two_pointers(s1, s2))

File: JavaScript/3-SlidingWindow/Problems/3-Permutation-in-String/Permutation-in-String.js
/*
Given two strings s1 and s2, return true if s2 contains a permutation of s1, or false otherwise.

In other words, return true if one of s1's permutations is a substring of s2.

Example 1:
Input: s1 = "ab", s2 = "eidbaooo"
Output: true
Explanation: s2 contains one permutation of s1 ("ba").

Example 2:
Input: s1 = "ab", s2 = "eidboaoo"
Output: false

Constraints:
1 <= s1.length, s2.length <= 104
s1 and s2 consist of lowercase English letters.
*/

/**
 * Approach 1: Brute Force (Checking all permutations of s1)
 *
 * Time Complexity: O(N! * M), where N is the length of s1 and M is the length of s2.  (N! to generate permutations, M to check if a permutation is a substring)
 * Space Complexity: O(N), for storing the current permutation.
 */
const checkInclusionBruteForce = (s1, s2) => {
  const getPermutations = (str) => {
    if (str.length === 1) {
      return [str];
    }
    const permutations = [];
    for (let i = 0; i < str.length; i++) {
      const char = str[i];
      const remainingChars = str.slice(0, i) + str.slice(i + 1);
      const subPermutations = getPermutations(remainingChars);
      for (const subPermutation of subPermutations) {
        permutations.push(char + subPermutation);
      }
    }
    return permutations;
  };

  const permutations = getPermutations(s1);
  for (const permutation of permutations) {
    if (s2.includes(permutation)) {
      return true;
    }
  }
  return false;
};



/**
 * Approach 2: Using HashMap (Character Frequency Counting)
 *
 * Time Complexity: O(N + M), where N is the length of s1 and M is the length of s2.
 * Space Complexity: O(N), for the hash map to store character frequencies of s1.
 */
const checkInclusionHashMap = (s1, s2) => {
  if (s1.length > s2.length) {
    return false;
  }

  const s1Freq = {};
  for (const char of s1) {
    s1Freq[char] = (s1Freq[char] || 0) + 1;
  }

  for (let i = 0; i <= s2.length - s1.length; i++) {
    const s2Freq = {};
    for (let j = 0; j < s1.length; j++) {
      const char = s2[i + j];
      s2Freq[char] = (s2Freq[char] || 0) + 1;
    }
    if (Object.keys(s1Freq).length === Object.keys(s2Freq).length) { // Important optimization
        let match = true;
        for (const char in s1Freq) {
          if (s1Freq[char] !== s2Freq[char]) {
            match = false;
            break;
          }
        }
        if (match) return true;
    }
  }
  return false;
};



/**
 * Approach 3: Sliding Window (Optimized HashMap)
 *
 * Time Complexity: O(N + M), where N is the length of s1 and M is the length of s2.
 * Space Complexity: O(N), for the hash map to store character frequencies of s1.
 */
const checkInclusionSlidingWindow = (s1, s2) => {
  if (s1.length > s2.length) {
    return false;
  }

  const s1Freq = {};
  for (const char of s1) {
    s1Freq[char] = (s1Freq[char] || 0) + 1;
  }

  const s2Freq = {};
  let left = 0;
  let right = 0;
  let matchCount = 0;

  while (right < s2.length) {
    const rightChar = s2[right];
    s2Freq[rightChar] = (s2Freq[rightChar] || 0) + 1;

    if (s1Freq[rightChar] && s2Freq[rightChar] <= s1Freq[rightChar]) {
      matchCount++;
    }

    if (right - left + 1 === s1.length) {
      if (matchCount === s1.length) {
        return true;
      }

      const leftChar = s2[left];
      if (s1Freq[leftChar] && s2Freq[leftChar] <= s1Freq[leftChar]) {
        matchCount--;
      }
      s2Freq[leftChar]--;
      left++;
    }
    right++;
  }
  return false;
};

/**
 * Approach 4: Using Arrays for Frequency Counting (Optimized for lowercase English letters)
 *
 * Time Complexity: O(N + M), where N is the length of s1 and M is the length of s2.
 * Space Complexity: O(1), because the array size is fixed (26 for lowercase English letters).
 */
const checkInclusionArray = (s1, s2) => {
  if (s1.length > s2.length) {
    return false;
  }

  const s1Freq = new Array(26).fill(0);
  const s2Freq = new Array(26).fill(0);

  for (const char of s1) {
    s1Freq[char.charCodeAt(0) - 'a'.charCodeAt(0)]++;
  }

  for (let i = 0; i < s1.length; i++) {
    s2Freq[s2[i].charCodeAt(0) - 'a'.charCodeAt(0)]++;
  }

  for (let i = 0; i <= s2.length - s1.length; i++) {
    let match = true;
    for (let j = 0; j < 26; j++) {
      if (s1Freq[j] !== s2Freq[j]) {
        match = false;
        break;
      }
    }
    if (match) {
      return true;
    }
    if (i < s2.length - s1.length) { // Important: Don't go out of bounds!
        s2Freq[s2[i].charCodeAt(0) - 'a'.charCodeAt(0)]--;
        s2Freq[s2[i + s1.length].charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }

  }
  return false;
};

/**
 * Approach 5: Sliding Window with Array (Optimized for lowercase English letters)
 *
 * Time Complexity: O(N + M), where N is the length of s1 and M is the length of s2.
 * Space Complexity: O(1), because the array size is fixed (26 for lowercase English letters).
 */
const checkInclusionSlidingWindowArray = (s1, s2) => {
    if (s1.length > s2.length) return false;

    const s1Freq = new Array(26).fill(0);
    const s2Freq = new Array(26).fill(0);

    for (let i = 0; i < s1.length; i++) {
        s1Freq[s1[i].charCodeAt(0) - 'a'.charCodeAt(0)]++;
        s2Freq[s2[i].charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }

    let matches = 0;
    for (let i = 0; i < 26; i++) {
        if (s1Freq[i] === s2Freq[i]) {
            matches++;
        }
    }

    let left = 0;
    for (let right = s1.length; right < s2.length; right++) {
        if (matches === 26) return true;

        let rightCharIndex = s2[right].charCodeAt(0) - 'a'.charCodeAt(0);
        let leftCharIndex = s2[left].charCodeAt(0) - 'a'.charCodeAt(0);

        s2Freq[rightCharIndex]++;
        if (s2Freq[rightCharIndex] === s1Freq[rightCharIndex]) {
            matches++;
        } else if (s2Freq[rightCharIndex] === s1Freq[rightCharIndex] + 1) {
            matches--;
        }

        s2Freq[leftCharIndex]--;
        if (s2Freq[leftCharIndex] === s1Freq[leftCharIndex]) {
            matches++;
        } else if (s2Freq[leftCharIndex] === s1Freq[leftCharIndex] - 1) {
            matches--;
        }
        left++;
    }
    return matches === 26;
};



// Example Usage and Verification
const s1 = "ab";
const s2 = "eidbaooo";
const s3 = "eidboaoo";

console.log("Input:");
console.log("s1:", s1);
console.log("s2:", s2);
console.log("s3:", s3);
console.log("\nOutput:");

// Brute Force
console.log("\nApproach 1: Brute Force");
console.log("s2 contains permutation of s1:", checkInclusionBruteForce(s1, s2)); // true
console.log("s3 contains permutation of s1:", checkInclusionBruteForce(s1, s3)); // false

// HashMap
console.log("\nApproach 2: Using HashMap");
console.log("s2 contains permutation of s1:", checkInclusionHashMap(s1, s2)); // true
console.log("s3 contains permutation of s1:", checkInclusionHashMap(s1, s3)); // false

// Sliding Window
console.log("\nApproach 3: Sliding Window");
console.log("s2 contains permutation of s1:", checkInclusionSlidingWindow(s1, s2)); // true
console.log("s3 contains permutation of s1:", checkInclusionSlidingWindow(s1, s3)); // false

// Array
console.log("\nApproach 4: Using Arrays");
console.log("s2 contains permutation of s1:", checkInclusionArray(s1, s2)); // true
console.log("s3 contains permutation of s1:", checkInclusionArray(s1, s3)); // false

// Sliding Window Array
console.log("\nApproach 5: Sliding Window with Array");
console.log("s2 contains permutation of s1:", checkInclusionSlidingWindowArray(s1, s2)); // true
console.log("s3 contains permutation of s1:", checkInclusionSlidingWindowArray(s1, s3)); // false
